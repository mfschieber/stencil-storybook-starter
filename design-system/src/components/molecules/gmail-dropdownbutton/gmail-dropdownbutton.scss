:host {
  gmail-button:hover {
    width: 100%;
    background: rgb(226, 222, 222);
  }
  position: relative;
  display: flex;
  width: fit-content;
  height: fit-content;
  .backdrop {
    position: fixed;
    top: 0;
    left: 0;
    z-index: 10;
    width: 100%;
    height: 100vh;
    pointer-events: all;

    &.hidden {
      pointer-events: none;
    }
  }

  .dropdown {
    position: absolute;
    top: 100%;
    left: 50%;
    z-index: 100;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    width: fit-content;
    height: fit-content;
    padding-top: 8px;
    padding-right: 0px;
    padding-bottom: 8px;
    padding-left: 0px;
    white-space: nowrap;
    background: white;
    border-radius: 1%;
    box-shadow: 0px 0px 2px grey;
    transform: translateY(0) translateX(0);
    opacity: 1;
    pointer-events: all;

    &.hidden {
      transform: translateY(-10px) translateX(-5px);
      opacity: 0;
      pointer-events: none;
    }
  }
}
