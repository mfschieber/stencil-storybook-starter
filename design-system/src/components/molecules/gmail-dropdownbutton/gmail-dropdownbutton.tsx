import { Component, ComponentInterface, h, Host, Method, Prop, State } from '@stencil/core';

import { Identifier } from '../../shared/identifier';

@Component({
  tag: 'gmail-dropdownbutton',
  styleUrl: 'gmail-dropdownbutton.scss',
  shadow: true,
})
export class GmailDropdownbutton implements ComponentInterface {
  @Prop() size?: 'large' | 'medium' | 'small' | 'xsmall' = 'medium';
  @Prop() type?: 'normal' | 'bold' | 'italic' = 'normal';
  @Prop() identifier?: Identifier = 'button';
  @Prop() items?: string[];

  @State() isOpen = false;

  logobuttonClickedHandler(event: CustomEvent<Identifier>) {
    if (event.detail === 'dropdownMenuButton') {
      this.toggle();
    } else {
      this.close();
    }
  }

  @Method()
  async open() {
    this.isOpen = true;
  }

  @Method()
  async close() {
    this.isOpen = false;
  }

  @Method()
  async toggle() {
    this.isOpen = !this.isOpen;
  }

  render() {
    return (
      <Host>
        <div class={`backdrop ${this.isOpen ? '' : 'hidden'}`} onClick={() => this.close()} />
        <gmail-logobutton
          class="dropdownbutton"
          size={this.size}
          type="dropdown"
          identifier="dropdownMenuButton"
          onLogobuttonClicked={(ev) => this.logobuttonClickedHandler(ev)}
        ></gmail-logobutton>
        <div class={`dropdown ${this.isOpen ? '' : 'hidden'}`}>
          {this.items.map((item) => (
            <gmail-textbutton size={this.size} content={item} type={this.type}></gmail-textbutton>
          ))}
        </div>
      </Host>
    );
  }
}
