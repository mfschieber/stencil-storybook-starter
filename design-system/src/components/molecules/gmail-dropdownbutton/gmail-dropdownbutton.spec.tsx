import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';

import { GmailDropdownbutton } from './gmail-dropdownbutton';

describe('ATOMS/gmail-dropdownbutton', () => {
  let page;

  it('should render', async () => {
    page = await newSpecPage({
      components: [GmailDropdownbutton],
      template: () => <gmail-dropdownbutton></gmail-dropdownbutton>,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should fire a click event', async () => {
    page = await newSpecPage({
      components: [GmailDropdownbutton],
      html: `<gmail-dropdownbutton></gmail-dropdownbutton>`,
    });

    const onClick = jest.fn();
    page.root.addEventListener('click', onClick);
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.root.click();
    expect(onClick).toHaveBeenCalled();
    expect(page.root).toMatchSnapshot();
  });
});
