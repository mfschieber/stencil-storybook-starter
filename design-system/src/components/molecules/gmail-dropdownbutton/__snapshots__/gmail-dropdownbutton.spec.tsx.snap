// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ATOMS/gmail-dropdownbutton should render 1`] = `
<gmail-dropdownbutton>
  <mock:shadow-root>
    <div class="backdrop hidden"></div>
    <gmail-logobutton class="dropdownbutton" identifier="dropdownMenuButton" size="medium" type="dropdown"></gmail-logobutton>
    <div class="dropdown hidden"></div>
  </mock:shadow-root>
</gmail-dropdownbutton>
`;
